sim4rec/response/nn_response.py

```diff
@@ -0,0 +1,204 @@
+        with open(os.path.join(path, "_item_indexer.pkl"), "wb") as f:
+            pickle.dump(self.item_indexer, f, pickle.HIGHEST_PROTOCOL)
+        with open(os.path.join(path, "_user_indexer.pkl"), "wb") as f:
+            pickle.dump(self.user_indexer, f, pickle.HIGHEST_PROTOCOL)
+        with open(os.path.join(path, "_params.pkl"), "wb") as f:
+            pickle.dump(
+                {
+                    "outputCol": self.outputCol,
+                    "log_dir": self.log_dir,
+                    "hist_data_dir": self.hist_data_dir,
+                },
+                f,
+                pickle.HIGHEST_PROTOCOL,
+            )
+
+    def _transform(self, new_recs):
+        """
+        Predict responses for given dataframe with recommendations.
+
+        :param dataframe: new recommendations.
+        """
+
+        def predict_udf(df):
+            # if not do this, something unstable happens to the Method Resolution Order
+            from .nn_utils.datasets import PandasRecommendationData
+
+            dataset = PandasRecommendationData(
+                log=df,
+                item_indexer=self.item_indexer,
+                user_indexer=self.user_indexer,
+            )
+
+            # replacing clicks in datset with predicted
+            dataset = self.backbone_response_model.transform(dataset=dataset)
+
+            return dataset._log[SIM_LOG_COLS]
+
+        spark = new_recs.sql_ctx.sparkSession
+
+        # read the historical data
+        hist_data = spark.read.schema(SIM_LOG_SCHEMA).parquet(self.hist_data_dir)
+        if not hist_data:
+            print("Warning: the historical data is empty")
+            hist_data = spark.createDataFrame([], schema=SIM_LOG_SCHEMA)
+        # filter users whom we don't need
+        hist_data = hist_data.join(new_recs, on="user_idx", how="inner").select(
+            hist_data["*"]
+        )
+
+        # read the updated simulator log
+        simlog = spark.read.schema(SIM_LOG_SCHEMA).parquet(self.log_dir)
+        if not simlog:
+            print("Warning: the simulator log is empty")
+            simlog = spark.createDataFrame([], schema=SIM_LOG_SCHEMA)
+        # filter users whom we don't need
+        simlog = simlog.join(new_recs, on="user_idx", how="inner").select(simlog["*"])
+
+        NEW_ITER_NO = 9999999
+
+        # since all the historical records are older than simulated by design,
+        # and new slates are newer than simulated, i can simply concat it
+        combined_data = hist_data.unionByName(simlog).unionByName(
+            new_recs.withColumn("response_proba", sf.lit(0.0))
+            .withColumn("response", sf.lit(0.0))
+            .withColumn(
+                "__iter",
+                sf.lit(
+                    NEW_ITER_NO
+                ),  # this is just a large number, TODO: add correct "__iter" field to sim4rec.sample_responses to avoid this constants
+            )
+        )
+
+        # not very optimal way, it makes one worker to
```

Comment:
need to discuss. you batch id should not influence the partitioning.  one partition != one batch and the users are grouped to batches within one partition. do not now how to implement it for now.

Reply:
BatchID won't influence the partition, because each batch must consist of the whole interaction history of a specific group of users. I